### PR TITLE
[PATCH v6] API: Crypto API additions

### PR DESCRIPTION
**Pull request by JannePeltonen**
api: crypto: clarify packet ordering of async crypto ops
api: crypto: add null_crypto crypto op flag for skipping crypto
api: crypto: allow both basic and out-of-place ops in one session

### CODE REVIEW

File: include/odp/api/spec/crypto.h [hunk 1]
```diff
@@ -296,6 +296,17 @@ int odp_crypto_op(const odp_packet_t pkt_in[],
  * All arrays should be of num_pkt size, except that pkt_out parameter
```
Comment (JereLeppanen) [inline include/odp/api/spec/crypto.h:302]: Commit comment: "between session" -> "between sessions".

File: include/odp/api/spec/crypto.h [hunk 2]
```diff
@@ -296,6 +296,17 @@ int odp_crypto_op(const odp_packet_t pkt_in[],
...
  * is ignored when the crypto operation type is ODP_CRYPTO_OP_TYPE_BASIC.
  *
+ * From packet ordering perspective this function behaves as if each input
+ * packet was enqueued to a crypto session specific ODP queue in the order
+ * the packets appear in the parameter array. The conceptial session input
```
Comment (JereLeppanen) [inline include/odp/api/spec/crypto.h:307]: Typo: "conceptial".

File: include/odp/api/spec/crypto_types.h [hunk 3]
```diff
@@ -611,6 +611,20 @@ typedef struct odp_crypto_session_param_t {
...
+	 *
+	 *  When this flag is true, the null_crypto flag of crypto operation
+	 *  parameters can be set to request skipping of ciphering and
+	 *  authentication of a packet regardless of session configuration.
+	 *  This may be useful for preserving packet order between packets
+	 *  that require crypto processing and packets that do not.
+	 *
+	 *  This flag cannot be set when op_mode is ODP_CRYPTO_SYNC.
+	 *
+	 *  The default value is false,
```
Comment (JereLeppanen) [inline include/odp/api/spec/crypto_types.h:639]: "," -> "."

File: include/odp/api/spec/crypto_types.h [hunk 4]
```diff
@@ -801,6 +815,14 @@ typedef struct odp_crypto_packet_op_param_t {
 	 */
 	int32_t dst_offset_shift;
 
+	/** Use null crypto algorithms
+	 *
+	 * Process packet using the null cipher and null auth algorithm
```
Comment (anoobj) [inline include/odp/api/spec/crypto_types.h:835]: Does this mean memcpy need to be done based on ranges?

File: include/odp/api/spec/crypto.h (hunk 1 above)
Comment (JannePeltonen) [reply include/odp/api/spec/crypto.h:302]: v2

File: include/odp/api/spec/crypto.h (hunk 2 above)
Comment (JannePeltonen) [reply include/odp/api/spec/crypto.h:307]: v2

File: include/odp/api/spec/crypto_types.h (hunk 3 above)
Comment (JannePeltonen) [reply include/odp/api/spec/crypto_types.h:639]: v2

**Comment by JannePeltonen** [comment]
v2: fixed typos and reworded one sentence (...f.lag cannot be set... --> ...flag must be set false...).

**Comment by JannePeltonen** [comment]
v3: added validation tests and implementation

**Review by psavol — APPROVED** [review]

**Comment by JannePeltonen** [comment]
v4: rebased

**Comment by anoobj** [comment]
ODP spec changes look good.

Reviewed-by: Anoob Joseph <anoobj@marvell.com>

**Pull request merged**